simulation/src/main/kotlin/it/unibo/collektive/examples/exercises/NetworkDiameter.kt

```diff
@@ -0,0 +1,79 @@
+package it.unibo.collektive.examples.exercises
+
+import it.unibo.collektive.aggregate.api.Aggregate
+import it.unibo.collektive.aggregate.api.Aggregate.Companion.neighboring
+import it.unibo.collektive.alchemist.device.sensors.EnvironmentVariables
+import it.unibo.alchemist.collektive.device.DistanceSensor
+import it.unibo.collektive.examples.exercises.distanceToSource
+import it.unibo.collektive.examples.channel.broadcast
+import it.unibo.collektive.field.operations.max
+import it.unibo.collektive.field.operations.min
+import it.unibo.collektive.field.operations.maxBy
+import it.unibo.collektive.stdlib.spreading.distanceTo
+
+/**
+ * 3) Calculate in the source an estimate of the true diameter of the network (the maximum distance of a device in the network).
+ * 4) Broadcast the diameter to every node in the network.
+*/ 
+
+fun Aggregate<Int>.networkDiameter(environment: EnvironmentVariables, distanceSensor: DistanceSensor): Int {
+    // Individuate source and calculate distance to source from the previous exercises
+    distanceToSource(environment)
+
+    // Calculate the distance of the local minimum for each neighborhood field
+    val distance: Int = environment["distanceToSource"]
+
+    // Identifies the node with the maximum number of hops to the source
+    val maxHopToSource = neighboring(distance).max(distance)
+
+    // Identifies the node furthest from the source, i.e. the most peripheral node
+    environment["furthest"] = isMaxValue(maxHopToSource, environment["distanceToSource"])
+
+    // Calculate distance to furthest node from the source in the network
+    val distanceToFurthest = distanceToFurthest(environment["furthest"])
+
+    // Identifies the node with the maximum number of hops corresponding to the diameter of the entire network
+    val flagNodeWithMaxHopToFurthest = isMaxValue(environment["distanceToFurthest"])
+
+    environment["distanceToFurthest"] = distanceToFurthest
+
+    // Only the identified node will broadcast its maximum number of hops to the entire network.
+    var broadcastMessage = broadcast(distanceSensor, from = flagNodeWithMaxHopToFurthest, payload = environment["distanceToFurthest"]).toInt()
+
+    if(distanceToFurthest <= broadcastMessage){
+        environment["networkDiameter"] = broadcastMessage
+    }else{
+        environment["networkDiameter"] = distanceToFurthest
+    }
+
+    environment["broadcastMessagePayload"] = broadcastMessage
+
+    return localId
```

Comment:
As a general rule: avoid setting into the environment values that will not be used.